plugins/pastefromword/filter/default.js

```diff
@@ -1541,34 +1541,140 @@
 			wholeImages = rtfContent.match( rePicture );
+
 			if ( !wholeImages ) {
 				return ret;
 			}
 
 			for ( var i = 0; i < wholeImages.length; i++ ) {
-				if ( rePictureHeader.test( wholeImages[ i ] ) ) {
-					if ( wholeImages[ i ].indexOf( '\\pngblip' ) !== -1 ) {
+				var currentImage = wholeImages[ i ];
+
+				if ( rePictureHeader.test( currentImage ) ) {
+					var id = getImageId( currentImage ),
+						imageDataIndex = CKEDITOR.tools.array.indexOf( ret, function( image ) {
+							return image.id === id;
+						} ),
+						isAlreadyExtracted = imageDataIndex !== -1 && ret[ imageDataIndex ].hex,
+						isWordArtShape = currentImage.indexOf( '\\defshp' ) !== -1,
+						isNonWordImage = currentImage.indexOf( '\\nonshppict' ) !== -1;
+
+					// This image is already extracted, it's WordArt shape or it's a non-Word version.
+					if ( isAlreadyExtracted || isWordArtShape || isNonWordImage ) {
+						continue;
+					}
+
+					if ( currentImage.indexOf( '\\pngblip' ) !== -1 ) {
 						imageType = 'image/png';
-					} else if ( wholeImages[ i ].indexOf( '\\jpegblip' ) !== -1 ) {
+					} else if ( currentImage.indexOf( '\\jpegblip' ) !== -1 ) {
 						imageType = 'image/jpeg';
 					} else {
-						continue;
+						imageType = 'unknown';
 					}
 
-					ret.push( {
-						hex: imageType ? wholeImages[ i ].replace( rePictureHeader, '' ).replace( /[^\da-fA-F]/g, '' ) : null,
+					var newImageData = {
+						id: id,
+						hex: imageType !== 'unknown' ?
+							currentImage.replace( rePictureHeader, '' ).replace( /[^\da-fA-F]/g, '' ) : null,
 						type: imageType
-					} );
+					};
+
+					if ( imageDataIndex !== -1 ) {
+						ret.splice( imageDataIndex, 1, newImageData );
+					} else {
+						ret.push( newImageData );
+					}
 				}
 			}
 
 			return ret;
+
+			function getImageId( image ) {
+				var blipUidRegex = /\\blipuid (\w+)\}/,
+					blipTagRegex = /\\bliptag(-?\d+)/,
+					blipUidMatch = image.match( blipUidRegex ),
+					blipTagMatch = image.match( blipTagRegex );
+
+				if ( blipUidMatch ) {
+					return blipUidMatch[ 1 ];
+				} else if ( blipTagMatch ) {
+					return blipTagMatch[ 1 ];
+				}
+
+				return null;
+			}
+
+			function removeHeadersAndFooters( rtfContent ) {
+				var startRegex = /\{\\(?:header|footer)[lrf]?/,
+					current;
+
+				while ( current = rtfContent.match( startRegex ) ) {
+					var precedingContent = rtfContent.substring( 0, current.index - 1 ),
+						contentToParse = rtfContent.substring( current.index ),
+						parsedContent = removeMatchedGroup( contentToParse );
+
+					rtfContent = precedingContent + parsedContent;
+
+					startRegex.lastIndex = 0;
+				}
+
+				return rtfContent;
+			}
+
+			// This function is in fact a very primitive RTF parser.
+			// It iterates over RTF content and search for the last } in the group
+			// by keeping track of how many elements are open using a stack-like method.
+			function removeMatchedGroup( content ) {
```

Comment:
And this method could be changed into `findGroup( content )` or `findRootGroup( rtfContent )`? Since it finds the biggest group in the given content from what I see? It will return group index and content and then can be used by `removeGroups()`, WDYT?